presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,280 @@
+        MaterializedViewInfo materializedViewInfo = new MaterializedViewInfo();
+        MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
+        materializedViewInformationExtractor.process(materializedViewQuery, materializedViewInfo);
+        this.materializedViewInfo = materializedViewInfo;
+
+        // If materialized view has a LIMIT clause, then it would reject the rewrite optimization
+        if (materializedViewInfo.isLimitClausePresented()) {
+            isRewritable = false;
+        }
+    }
+
+    public Node rewrite(Node node)
+    {
+        if (!isRewritable) {
+            return node;
+        }
+        return process(node);
+    }
+
+    @Override
+    protected Node visitNode(Node node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected Node visitQuery(Query node, Void context)
+    {
+        QueryBody rewrittenQueryBody = (QueryBody) process(node.getQueryBody(), context);
+        if (!isRewritable) {
+            return node;
+        }
+        return new Query(
+                node.getWith(),
+                rewrittenQueryBody,
+                node.getOrderBy(),
+                node.getOffset(),
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
+    {
+        // Each of the process could potentially set isRewritable to false. So return as early as possible is an potential optimization
+        // TODO: Reject the rewrite optimization as soon as possible (optimization opportunity)
+
+        // Process from and select clause before all others to check if query is rewritable and acquire alias relation if any
+        Optional<Relation> rewrittenFrom = node.getFrom().map(from -> (Relation) process(from, context));
+        Select rewrittenSelect = (Select) process(node.getSelect(), context);
+
+        // In WHERE clause validation process, validator only allows two conditions for rewrite so far:
+        // 1. MV's definition query has no WHERE clause
+        // 2. MV's WHERE clause is the exact match of base query's WHERE clause
+        // TODO: Improve the implication logic in the validator so that it could handle more eligible conditions
+        if (materializedViewInfo.getMaterializedViewDefinitionWhere().isPresent() && !materializedViewInfo.getMaterializedViewDefinitionWhere().get().equals(node.getWhere().orElse(null))) {
+            isRewritable = false;
+            return node;
+        }
+        Optional<Expression> rewrittenWhere = node.getWhere().map(where -> (Expression) process(where, context));
+
+        // If materialized view contains groupBy clause but base query does not, then it rejects rewrite optimization.
+        if (materializedViewInfo.getMaterializedViewDefinitionGroupBy().isPresent() && !node.getGroupBy().isPresent()) {
+            isRewritable = false;
```

Comment:
I am wondering why do we need to handle this here? This can be handled in the visitGroupBy only, right?

Reply:
@jainxrohit The optimizer will only process visitGroupBy if there exists a GroupBy in the base query. If not, then it will skip the visitGroupBy.

Here I am only trying to eliminate case that does not have groupBy when the MV has groupBy.